numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2923,8 +2923,36 @@ static PyObject *
 
     /* TODO: include type name in error message, which is not @name@ */
     PyObject *obj = NULL;
+#if defined(_@TYPE@_IS_CDOUBLE) || defined(_@TYPE@_IS_CFLOAT)
+    static char *kwnames[] = {"", "", NULL};  /* positional-only */
+    static char *parse_error = "Could not convert arguments into a complex scalar. '%R' given.";
+    PyObject *real_obj = NULL;
+    PyObject *imag_obj = NULL;
+    if (PyArg_ParseTupleAndKeywords(args, kwds, "|OO", kwnames, &real_obj, &imag_obj)) {
+        if (imag_obj == NULL) {
+            obj = real_obj;
+            Py_XINCREF(obj);
+        }
+        else if (PyNumber_Check(real_obj) && PyNumber_Check(imag_obj)) {
+            if (PyComplex_Check(real_obj) || PyComplex_Check(imag_obj)) {
+                PyErr_Format(PyExc_TypeError, parse_error, args);
+                return NULL;
+            }
+            obj = PyComplex_Type.tp_new(&PyComplex_Type, args, NULL);
+            if (obj == NULL) {
+                return NULL;
+            }
```

Comment:
While annoying, duplicating the scalar creation may be easier here?  Something like this (but I am not sure the cast is actually valid, may have to cast real/imag separately. 
```
npy_@name@ val = PyComplex_AsCComplex(obj);  // does that work, since its struct-like
// Need to fetch the typecode.  Note that this can't actually fail
// (unlike the code below makes you believe).
PyObject *robj = PyArray_Scalar(&val, typecode, NULL);
Py_DECREF(py_complex);
return robj;
```

Reply:
Unfortunately Numpy at the moment allows any array-like to be passed to the scalar initializers (and there are tests for it), which means I'd first need to re-implement the array-like coercion logic in `PyArray_FromAny` to get this to work.

```
In [2]: np.complex128([1, 2, 3, 4])
Out[2]: array([1.+0.j, 2.+0.j, 3.+0.j, 4.+0.j])

In [3]: np.complex128(b"1+4j")
Out[3]: (1+4j)
```

I think it's better to just avoid the early return and re-use the logic in the rest of this function that goes through `PyArray_FromAny`.